src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp

```diff
@@ -1609,12 +1594,59 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
 			_att_sp.pitch_body = radians(_takeoff_pitch_min.get());
 		}
 	}
+
+	if (_launch_detection_state != LAUNCHDETECTION_RES_DETECTED_ENABLEMOTORS && !_runway_takeoff.runwayTakeoffEnabled()) {
+
+		/* making sure again that the correct thrust is used,
+		 * without depending on library calls for safety reasons.
+		   the pre-takeoff throttle and the idle throttle normally map to the same parameter. */
+		_att_sp.thrust_body[0] = _param_fw_thr_idle.get();
+
+	} else if (_runway_takeoff.runwayTakeoffEnabled()) {
+
+		_att_sp.thrust_body[0] = _runway_takeoff.getThrottle(now, get_tecs_thrust());
+
+	} else {
+		/* Copy thrust and pitch values from tecs */
+		if (_landed) {
+			// when we are landed state we want the motor to spin at idle speed
+			_att_sp.thrust_body[0] = min(_param_fw_thr_idle.get(), 1.f);
+
+		} else {
+			_att_sp.thrust_body[0] = get_tecs_thrust();
+		}
+	}
+
+	// decide when to use pitch setpoint from TECS because in some cases pitch
+	// setpoint is generated by other means
+	bool use_tecs_pitch = true;
+
+	// auto runway takeoff
+	use_tecs_pitch &= !(_launch_detection_state == LAUNCHDETECTION_RES_NONE || _runway_takeoff.runwayTakeoffEnabled());
+
+	if (use_tecs_pitch) {
+		_att_sp.pitch_body = get_tecs_pitch();
+	}
+
+	if (!_vehicle_status.in_transition_to_fw) {
+		publishLocalPositionSetpoint(pos_sp_curr);
+	}
 }
 
 void
-FixedwingPositionControl::control_auto_landing(const hrt_abstime &now, const float dt, const Vector2d &curr_pos,
+FixedwingPositionControl::control_auto_landing(const hrt_abstime &now, const Vector2d &curr_pos,
 		const Vector2f &ground_speed, const position_setpoint_s &pos_sp_prev, const position_setpoint_s &pos_sp_curr)
 {
+	const float dt = math::constrain((now - _control_position_last_called) * 1e-6f, 0.01f, 0.05f);
```

Comment:
does this appear multiple times? can encapsulate and additionally make the dt bounds constants?

Reply:
It does, but this is because each flight mode needs to keep track of the time steps separately. What do you mean by encapsulation? Where would you suggest that we encapsulate the time tracking?

Reply:
ah now i see it has a different time tracker for each one. ughh. ok. still would be nice to reduce magic numbers on the constraints if they are the same for each.